File: backend/pkgs/printer/brother_raster.go
package printer

import (
	"bytes"
	"context"
	"fmt"
	"image"
	"image/png"
	"net"
	"time"
)

// BrotherRasterClient implements PrinterClient for Brother QL label printers
// using the Brother raster protocol over raw TCP (port 9100)
type BrotherRasterClient struct {
	address string
	port    string
	media   BrotherMediaType
}

// BrotherPrintOptions contains options for a Brother print job
type BrotherPrintOptions struct {
	MediaType     string  // Media type identifier (e.g., "DK-22251")
	LabelLengthMM float64 // Label length in mm (for continuous rolls)
}

// NewBrotherRasterClient creates a new Brother raster client
// Address should be the IP or hostname of the printer
func NewBrotherRasterClient(address string) (*BrotherRasterClient, error) {
	return &BrotherRasterClient{
		address: address,
		port:    "9100",
		media:   BrotherMediaTypes["DK-22251"], // Default to DK-22251
	}, nil
}

// SetMediaType configures the media type for the printer
func (c *BrotherRasterClient) SetMediaType(mediaType string) error {
	media, ok := GetBrotherMedia(mediaType)
	if !ok {
		return fmt.Errorf("unknown media type: %s", mediaType)
	}
	c.media = media
	return nil
}

// GetPrinterInfo returns information about the printer
func (c *BrotherRasterClient) GetPrinterInfo(ctx context.Context) (*PrinterInfo, error) {
	return &PrinterInfo{
		Name:  "Brother QL",
		Make:  "Brother",
		Model: "QL-820NWB",
		State: StatusOnline,
	}, nil
}

// Print sends a document to the printer
func (c *BrotherRasterClient) Print(ctx context.Context, job *PrintJob) (*PrintResult, error) {
	// Decode PNG image
	img, err := png.Decode(bytes.NewReader(job.Data))
	if err != nil {
		return nil, fmt.Errorf("failed to decode PNG: %w", err)
	}

	// Convert to Brother raster format
	rasterData := c.convertToRaster(img)

	// Connect to printer
	conn, err := net.DialTimeout("tcp", fmt.Sprintf("%s:%s", c.address, c.port), 10*time.Second)
	if err != nil {
		return nil, fmt.Errorf("failed to connect to printer: %w", err)
	}
	defer func() { _ = conn.Close() }()

	// Set write deadline
	if err := conn.SetWriteDeadline(time.Now().Add(30 * time.Second)); err != nil {
		return nil, fmt.Errorf("failed to set write deadline: %w", err)
	}

	// Send data
	_, err = conn.Write(rasterData)
	if err != nil {
		return nil, fmt.Errorf("failed to send data: %w", err)
	}

	return &PrintResult{
		Success: true,
		Message: "Print job sent successfully",
	}, nil
}

// convertToRaster converts an image to Brother raster format
func (c *BrotherRasterClient) convertToRaster(img image.Image) []byte {
	bounds := img.Bounds()
	imgWidth := bounds.Dx()
	imgHeight := bounds.Dy()

	// Brother QL uses 90 bytes per line (720 dots) for 62mm, less for narrower
	bytesPerLine := 90
	printableWidth := c.media.PrintableDotsWidth
	if printableWidth == 0 {
		printableWidth = 696 // Default for 62mm
	}

	// Feed lines to add before and after content (about 3mm at 300 DPI)
	feedLines := 35
	// Total raster lines = feed + image + feed
	totalRasterLines := feedLines + imgHeight + feedLines

	var buf bytes.Buffer

	// 1. Invalidate - 200 null bytes
	buf.Write(make([]byte, 200))

	// 2. Initialize - ESC @
	buf.Write([]byte{0x1B, 0x40})

	// 3. Switch to raster mode - ESC i a 1
	buf.Write([]byte{0x1B, 0x69, 0x61, 0x01})

	// 4. Print info - ESC i z
	// Flags byte: 0x80 = quality, 0x08 = length valid, 0x04 = width valid, 0x02 = type valid
	flags := byte(0x80 | 0x08 | 0x04 | 0x02) // 0x8E
	buf.Write([]byte{0x1B, 0x69, 0x7A})
	buf.WriteByte(flags)
	buf.WriteByte(c.media.MediaCode) // Media type (0x0A=continuous, 0x0B=die-cut)
	buf.WriteByte(byte(c.media.WidthMM))
	buf.WriteByte(byte(c.media.LengthMM)) // Length in mm (0 for continuous)
	// Number of raster lines (4 bytes, little endian) - includes feed lines
	buf.WriteByte(byte(totalRasterLines & 0xFF))
	buf.WriteByte(byte((totalRasterLines >> 8) & 0xFF))
	buf.WriteByte(byte((totalRasterLines >> 16) & 0xFF))
	buf.WriteByte(byte((totalRasterLines >> 24) & 0xFF))
	buf.WriteByte(0) // First page (0 = first page only)
	buf.WriteByte(0) // Reserved

	// 5. Expanded mode - ESC i K
	// Bit 3 (0x08) = cut at end
	// Bit 0 (0x01) = two-color mode (for DK-22251, etc.)
	expandedFlags := byte(0x08)
	if c.media.TwoColor {
		expandedFlags |= 0x01
	}
	buf.Write([]byte{0x1B, 0x69, 0x4B, expandedFlags})

	// 6. Auto cut - ESC i M
	// 0x40 = auto cut enabled
	buf.Write([]byte{0x1B, 0x69, 0x4D, 0x40})

	// 7. Margin - ESC i d
	// Set margin to 35 dots (standard for continuous rolls)
	marginDots := uint16(35)
	buf.Write([]byte{0x1B, 0x69, 0x64, byte(marginDots & 0xFF), byte(marginDots >> 8)})

	// 8. Compression mode - M 0 (no compression)
	buf.Write([]byte{0x4D, 0x00})

	// 9. Raster data
	// Center the image in the printable area
	offsetDots := (printableWidth - imgWidth) / 2
	if offsetDots < 0 {
		offsetDots = 0
	}

	// Helper to write a raster line (blank or with data)
	// redData is optional - if nil, an empty red plane is sent in two-color mode
	writeRasterLine := func(blackData []byte, redData []byte) {
		if c.media.TwoColor {
			// Two-color mode: send BOTH planes for each line
			// First the black plane (0x01)
			buf.Write([]byte{0x77, 0x01, byte(bytesPerLine)})
			buf.Write(blackData)
			// Then the red plane (0x02) - empty if not provided
			buf.Write([]byte{0x77, 0x02, byte(bytesPerLine)})
			if redData != nil {
				buf.Write(redData)
			} else {
				buf.Write(make([]byte, bytesPerLine)) // Empty red plane
			}
		} else {
			// Standard mode: use 0x67 0x00
			buf.Write([]byte{0x67, 0x00, byte(bytesPerLine)})
			buf.Write(blackData)
		}
	}

	// Add leading blank lines for proper feed (about 3mm worth at 300 DPI = ~35 lines)
	blankLine := make([]byte, bytesPerLine)
	for i := 0; i < feedLines; i++ {
		writeRasterLine(blankLine, nil)
	}

	// Write the actual image data
	// Note: Brother QL printers require horizontally mirrored data
	for y := bounds.Min.Y; y < bounds.Max.Y; y++ {
		blackData := make([]byte, bytesPerLine)
		var redData []byte
		if c.media.TwoColor {
			redData = make([]byte, bytesPerLine)
		}

		for x := bounds.Min.X; x < bounds.Max.X; x++ {
			rv, gv, bv, _ := img.At(x, y).RGBA()
			// Convert from 16-bit to 8-bit
			r8 := float64(rv >> 8)
			g8 := float64(gv >> 8)
			b8 := float64(bv >> 8)

			// Calculate grayscale value
			gray := 0.299*r8 + 0.587*g8 + 0.114*b8

			// Mirror horizontally: map x to (imgWidth - 1 - x)
			mirroredX := imgWidth - 1 - (x - bounds.Min.X)
			dotPos := offsetDots + mirroredX
			byteIdx := dotPos / 8
			bitIdx := 7 - (dotPos % 8)

			if byteIdx < 0 || byteIdx >= bytesPerLine {
				continue
			}

			if gray < 200 { // Not white (threshold for printing)
				if c.media.TwoColor {
					// For two-color mode, detect if pixel is "red"
					// Red pixels have high R, low G, low B
					isRed := r8 > 150 && g8 < 100 && b8 < 100
					if isRed {
						redData[byteIdx] |= (1 << bitIdx)
					} else {
						blackData[byteIdx] |= (1 << bitIdx)
					}
				} else if gray < 128 {
					// Single color mode: all dark pixels are black
					blackData[byteIdx] |= (1 << bitIdx)
				}
			}
		}

		writeRasterLine(blackData, redData)
	}

	// Add trailing blank lines for proper feed before cut (about 3mm = ~35 lines)
	for i := 0; i < feedLines; i++ {
		writeRasterLine(blankLine, nil)
	}

	// 10. Print command
	buf.WriteByte(0x1A)

	return buf.Bytes()
}

// GetJobStatus checks the status of a print job
// Brother raster protocol doesn't support job status queries
func (c *BrotherRasterClient) GetJobStatus(ctx context.Context, jobID int) (*JobStatus, error) {
	return &JobStatus{
		JobID:     jobID,
		State:     "completed",
		Completed: true,
	}, nil
}

// CancelJob cancels a pending print job
// Brother raster protocol doesn't support job cancellation
func (c *BrotherRasterClient) CancelJob(ctx context.Context, jobID int) error {
	return nil
}

File: frontend/lib/printing/protocols/brother-raster.ts
// Brother Raster Protocol implementation for QL series label printers
// Based on Brother QL series command reference

interface BrotherPrintOptions {
  width: number; // Label width in mm
  height?: number; // Label height in mm (for die-cut labels)
  copies: number;
  highQuality?: boolean;
  cutAtEnd?: boolean;
}

// Brother QL command constants
const ESC = 0x1b;
const NULL = 0x00;

// Media info for common Brother label sizes
const MEDIA_WIDTHS: Record<number, number> = {
  12: 106, // 12mm = 106 dots
  29: 306, // 29mm = 306 dots
  38: 413, // 38mm = 413 dots
  50: 554, // 50mm = 554 dots
  54: 590, // 54mm = 590 dots
  62: 696, // 62mm = 696 dots
  102: 1164, // 102mm = 1164 dots
};

function invalidate(): number[] {
  // Send 200 null bytes to clear any pending data
  return new Array(200).fill(NULL);
}

function initialize(): number[] {
  // ESC @ - Initialize printer
  return [ESC, 0x40];
}

function switchToRasterMode(): number[] {
  // ESC i a {n} - Switch to raster mode (n=1)
  return [ESC, 0x69, 0x61, 0x01];
}

function setMediaInfo(widthMm: number, heightMm?: number): number[] {
  // ESC i z {flags} {mediaType} {width} {height_low} {height_high} {pageCount_low} {pageCount_high} {startingPage} {0}
  const cmd = [ESC, 0x69, 0x7a];

  // Flags: bit 6 = quality, bit 7 = recover
  const flags = 0x86;
  cmd.push(flags);

  // Media type: 0x0A = continuous, 0x0B = die-cut
  const mediaType = heightMm ? 0x0b : 0x0a;
  cmd.push(mediaType);

  // Width (in mm, single byte)
  cmd.push(widthMm);

  // Height in mm (2 bytes, little endian)
  const height = heightMm || 0;
  cmd.push(height & 0xff);
  cmd.push((height >> 8) & 0xff);

  // Page count (2 bytes) - not used, set to 0
  cmd.push(0, 0);

  // Starting page - 0
  cmd.push(0);

  // Reserved - 0
  cmd.push(0);

  return cmd;
}

function setPrintMode(highQuality: boolean): number[] {
  // ESC i M {n} - Set print mode
  // n=0: Standard, n=1: High quality
  return [ESC, 0x69, 0x4d, highQuality ? 0x01 : 0x00];
}

function setCutOptions(cutAtEnd: boolean): number[] {
  // ESC i K {n} - Set cut options
  // bit 3: cut at end
  const flags = cutAtEnd ? 0x08 : 0x00;
  return [ESC, 0x69, 0x4b, flags];
}

function setCopies(copies: number): number[] {
  // ESC i A {n} - Set number of copies
  return [ESC, 0x69, 0x41, Math.min(copies, 255)];
}

function printCommand(): number[] {
  // Print and feed - end of page
  return [0x1a];
}

/**
 * Convert a PNG image to Brother raster format
 * For now, this creates a simple raster representation
 * In production, you'd want to use a proper image processing library
 */
function convertImageToRaster(_imageData: Uint8Array, widthMm: number): number[] {
  const rasterCommands: number[] = [];

  // Get the width in dots (300 DPI)
  const widthDots = MEDIA_WIDTHS[widthMm] || Math.round((widthMm * 300) / 25.4);
  const bytesPerLine = Math.ceil(widthDots / 8);

  // For a real implementation, we would:
  // 1. Decode the PNG image
  // 2. Convert to 1-bit black/white
  // 3. Generate raster lines

  // For now, we'll pass through the image data with raster line commands
  // This is a placeholder - real implementation would process the PNG

  // Each raster line: g {0x00} {0x5a} {length_low} {length_high} {data...}
  // The 'g' command sends a raster line

  // Create a simple test pattern (checkerboard) as placeholder
  // In production, this would be replaced with actual image conversion
  const heightDots = 200; // Placeholder height

  for (let y = 0; y < heightDots; y++) {
    // Raster line command
    rasterCommands.push(0x67, 0x00, 0x5a);

    // Length (2 bytes, little endian)
    rasterCommands.push(bytesPerLine & 0xff);
    rasterCommands.push((bytesPerLine >> 8) & 0xff);

    // Raster data - placeholder pattern
    for (let x = 0; x < bytesPerLine; x++) {
      // This would be actual image data in production
      rasterCommands.push(0x00);
    }
  }

  return rasterCommands;
}

/**
 * Brother Raster Protocol utilities for QL series label printers
 */
export const BrotherRasterProtocol = {
  /**
   * Create a complete print job for Brother QL printers
   * @param imageData PNG image data as Uint8Array
   * @param options Print options
   * @returns Uint8Array containing the complete print job
   */
  createPrintJob(imageData: Uint8Array, options: BrotherPrintOptions): Uint8Array {
    const commands: number[] = [];

    // 1. Invalidate - clear print buffer
    commands.push(...invalidate());

    // 2. Initialize - reset printer
    commands.push(...initialize());

    // 3. Switch to raster mode
    commands.push(...switchToRasterMode());

    // 4. Set media info
    commands.push(...setMediaInfo(options.width, options.height));

    // 5. Set print mode (quality)
    commands.push(...setPrintMode(options.highQuality ?? true));

    // 6. Set cut options
    commands.push(...setCutOptions(options.cutAtEnd ?? true));

    // 7. Set number of copies
    if (options.copies > 1) {
      commands.push(...setCopies(options.copies));
    }

    // 8. Convert image to raster data and add raster lines
    const rasterData = convertImageToRaster(imageData, options.width);
    commands.push(...rasterData);

    // 9. Print command
    commands.push(...printCommand());

    return new Uint8Array(commands);
  },

  /**
   * Parse printer status response
   */
  parseStatus(data: Uint8Array): {
    mediaWidth: number;
    mediaType: string;
    error: string | null;
  } {
    if (data.length < 32) {
      return { mediaWidth: 0, mediaType: "unknown", error: "Invalid status response" };
    }

    // Status byte locations in Brother QL response
    const errorInfo1 = data[8] ?? 0;
    const errorInfo2 = data[9] ?? 0;
    const mediaWidth = data[10] ?? 0;
    const mediaType = data[11] ?? 0;

    let error: string | null = null;

    // Check for errors
    if (errorInfo1 & 0x01) error = "No media";
    else if (errorInfo1 & 0x02) error = "End of media";
    else if (errorInfo1 & 0x04) error = "Tape cutter jam";
    else if (errorInfo1 & 0x10) error = "Transmission error";
    else if (errorInfo2 & 0x01) error = "Cover open";
    else if (errorInfo2 & 0x02) error = "Overheating";

    return {
      mediaWidth,
      mediaType: mediaType === 0x0a ? "continuous" : "die-cut",
      error,
    };
  },
};
